### PR TITLE
Rationalize how temporary directory works today

### PR DESCRIPTION
**Pull request by jaredpar**
Before fixing https://github.com/dotnet/roslyn/issues/65415 I needed to better rationalize how
`Path.GetTempPath` and temp paths in general worked in the compiler.
This change ended up addressing the following problems:

**Path.GetTempPath**

`Path.GetTempPath` should never be used in the compiler libraries.
That uses a number of environment variables, like working directory,
that can change during the lifetime of a build operation. That makes
compilation non-deterministic and can lead to sporadic errors. All
uses of `GetTempPath` need to come from the initial csc / vbc
invocation and be passed through as arguments.

This was solved by adding a new BannedSymbols file for compiler
libraries that banned `Path.GetTempPath` and working through the
remaining case I found.

This introduced one new error because `Path.GetTempPath` was being used
as a fall back in legacy signing if csc / vbc couldn't calculate a
temp path at invocation time. This is a "this should never actually
happen" problem but I added an error code as a precaution.

I did not force the temp directory to non-null for all compilations
because the vast majority of compilations don't actually need a temp
directory. It is only used in legacy signing which is very rare.

**Where is temp required**

It was unclear in our code base where a temp directory was actually
required for execution. To get a better handle on this, now and going
forward, I enabled NRT in files that handled temp directories. After
working through it there are only two cases where it's required:

1. When using legacy strong named signing.
2. When starting the compiler server.

For (2) it's required as a place to shadow copy analyzers /
generators. There is no reasonable fall back if the server can't
calculate a temp path. Further if the client can't calculate a temp path
then there is a very strong chance (basically guaranteed) that the
server can't either. Therefore I added a few places where we simply
don't even try to communicate with the compiler server if we can't find
a temporary directory. Basically fail fast vs. failing slow.

Once this change is in I can move forward with https://github.com/dotnet/roslyn/issues/65415. The impact of
that change will be much clearer after this goes through because it
changes both what we depend on and where shadow copy happens.

### CODE REVIEW

**Comment by RikkiGibson** [comment]
Since we don't want analyzers/generators doing I/O generally, it follows that we don't want them using temp directories for scratch space, etc., right?

**Comment by jaredpar** [comment]
> Since we don't want analyzers/generators doing I/O generally, it follows that we don't want them using temp directories for scratch space, etc., right?

Most definitely. If that's not o nthe banned list yet it needs to be.

File: src/Compilers/Core/Portable/Compilation.EmitStreamProvider.cs [hunk 1]
```diff
@@ -21,14 +21,17 @@ public abstract partial class Compilation
     {
         /// <summary>
         /// Abstraction that allows the caller to delay the creation of the <see cref="Stream"/> values 
-        /// until they are actually needed.
+        /// until they are actually needed. The <see cref="Stream"/> provided here is soley owned by
```
Comment (cston) [inline src/Compilers/Core/Portable/Compilation.EmitStreamProvider.cs:24]: >soley [](http://example.com/codeflow?start=86&length=5)

Typo? Consider simply removing the word.

File: .vscode/settings.json [hunk 2]
```diff
@@ -1,38 +0,0 @@
-{
```
Comment (RikkiGibson) [inline .vscode/settings.json:1]: I wasn't sure why this file was deleted, could you explain?

**Comment by cston** [comment]
>             private Stream? _streamToDispose;

This field is now duplicated in the base class. As an alternative, what if we remove the field from the base class and remove `CreateStream()` and change `GetOrCreateStream()` to the `abstract` method that is overridden in the two derived classes?

---
Refers to: src/Compilers/Core/Portable/CommandLine/CommonCompiler.CompilerEmitStreamProvider.cs:24 in eca53ed. [](commit_id = eca53eda2caddaee60731154569860ef46f0c8e6, deletion_comment = False)

File: src/Compilers/CSharp/Test/Emit/Emit/DesktopStrongNameProviderTests.cs [hunk 3]
```diff
@@ -21,14 +21,14 @@ public void RespectCustomTempPath()
         {
             var tempDir = Temp.CreateDirectory();
             var provider = new DesktopStrongNameProvider(tempPath: tempDir.Path);
-            Assert.Equal(tempDir.Path, provider.FileSystem.GetTempPath());
+            Assert.Equal(tempDir.Path, provider.FileSystem.GetSigningTempPath());
         }
 
         [Fact]
-        public void RespectDefaultTempPath()
+        public void RespectEmptyTempPath()
```
Comment (RikkiGibson) [inline src/Compilers/CSharp/Test/Emit/Emit/DesktopStrongNameProviderTests.cs:28]: I didn't understand how the new name of the test reflects the new assert.

File: src/Compilers/Shared/BuildClient.cs [hunk 4]
```diff
@@ -219,6 +217,13 @@ private int RunLocalCompilation(string[] arguments, BuildPaths buildPaths, TextW
                 return null;
             }
 
+            // A temporary directory is required for the server to startup hence we bail out early
+            // here if it's not available.
+            if (buildPaths.TempDirectory is null)
+            {
+                return null;
```
Comment (cston) [inline src/Compilers/Shared/BuildClient.cs:224]: >return null; [](http://example.com/codeflow?start=16&length=12)

Should we write a message to the log here?

File: src/Compilers/Shared/BuildServerConnection.cs [hunk 5]
```diff
@@ -13,6 +13,7 @@
 using System.Text;
 using System.Threading;
 using System.Threading.Tasks;
+using System.Web;
```
Comment (cston) [inline src/Compilers/Shared/BuildServerConnection.cs:16]: >System.Web [](http://example.com/codeflow?start=6&length=10)

Is this used?

File: src/Compilers/Core/Portable/Compilation.EmitStream.cs [hunk 6]
```diff
@@ -42,11 +42,16 @@ internal enum EmitStreamSignKind
         /// a temporary file must be created on disk (at the last possible moment), emitted to, signed on disk
         /// and then copied back to the original <see cref="Stream"/>. Only when legacy signing is enabled though.
         /// </summary>
+        /// <remarks>
+        /// The <see cref="Stream"/> returned here should not be disposed by the caller. It is either owned by
```
Comment (RikkiGibson) [inline src/Compilers/Core/Portable/Compilation.EmitStream.cs:46]: I find this language a little confusing. A type doesn't return anything. Was this remark meant to be on 'CreateStream'?

File: src/Compilers/Shared/BuildClient.cs [hunk 7]
```diff
@@ -219,6 +217,13 @@ private int RunLocalCompilation(string[] arguments, BuildPaths buildPaths, TextW
                 return null;
             }
 
+            // A temporary directory is required for the server to startup hence we bail out early
```
Comment (RikkiGibson) [inline src/Compilers/Shared/BuildClient.cs:220]: Is this a condition we will need to be able to diagnose in the wild? If so, should something be logged or thrown here?

It's not clear to me under what conditions the client can fail to get a temp path, and whether we should just keep going with a client-side compilation in that case. It feels like if this happens, it will just manifest as a mysterious increase in solution build time for the end user.

File: src/Compilers/Shared/BuildServerConnection.cs (hunk 5 above)
Comment (RikkiGibson) [inline src/Compilers/Shared/BuildServerConnection.cs:16]: This using looks unused.

File: src/Compilers/Shared/BuildServerConnection.cs [hunk 8]
```diff
@@ -588,51 +602,61 @@ internal static string GetClientMutexName(string pipeName)
...
 
         /// <summary>
-        /// Gets the value of the temporary path for the current environment assuming the working directory
-        /// is <paramref name="workingDir"/>.  This function must emulate <see cref="Path.GetTempPath"/> as 
-        /// closely as possible.
+        /// Gets the value of the temporary path for the provided environment settings. This behavior
+        /// is OS specific: 
+        ///   - On Windows it seeks to emulate Path.GetTempPath as closely as possible with 
+        ///     provided working directory.
+        ///   - On Unix it returns a directory under $XDG_RUNTIME_DIR if it is set using appropriate fall backs
```
Comment (RikkiGibson) [inline src/Compilers/Shared/BuildServerConnection.cs:609]: It looks like the implementation doesn't check this environment variable as described. Instead it's using `$TMPDIR` or `/tmp`. Did I misunderstand something?

File: src/Compilers/VisualBasic/Test/Emit/Attributes/EmitTestStrongNameProvider.vb [hunk 9]
```diff
@@ -20,7 +20,7 @@ Partial Public Class InternalsVisibleToAndStrongNameTests
     Public Sub BadInputStream()
         SigningTestHelpers.InstallKey()
         Dim thrownException = New IOException("This is a test IOException")
-        Dim testFileSystem = New TestStrongNameFileSystem() With {
+        Dim testFileSystem = New TestStrongNameFileSystem(Temp.CreateDirectory().Path) With {
```
Comment (RikkiGibson) [inline src/Compilers/VisualBasic/Test/Emit/Attributes/EmitTestStrongNameProvider.vb:23]: no need to dispose of the result of CreateDirectory, right?

File: src/Compilers/CSharp/Test/Emit2/Attributes/InternalsVisibleToAndStrongNameTests.cs [hunk 10]
```diff
@@ -178,6 +184,22 @@ public void PubKeyFromKeyFileAttribute_AssemblyKeyFileResolver_RelativeToCurrent
             Assert.True(ByteSequenceComparer.Equals(s_publicKey, comp.Assembly.Identity.PublicKey));
         }
 
+        [Fact]
+        public void PubKeyFromKeyFileAttribute_SigningTempPathNotAvailable()
+        {
+            var x = s_keyPairFile;
```
Comment (RikkiGibson) [inline src/Compilers/CSharp/Test/Emit2/Attributes/InternalsVisibleToAndStrongNameTests.cs:190]: Consider using a more descriptive variable name.

File: src/Compilers/CSharp/Test/Emit2/Attributes/InternalsVisibleToAndStrongNameTests.cs [hunk 11]
```diff
@@ -178,6 +184,22 @@ public void PubKeyFromKeyFileAttribute_AssemblyKeyFileResolver_RelativeToCurrent
             Assert.True(ByteSequenceComparer.Equals(s_publicKey, comp.Assembly.Identity.PublicKey));
         }
 
+        [Fact]
+        public void PubKeyFromKeyFileAttribute_SigningTempPathNotAvailable()
```
Comment (RikkiGibson) [inline src/Compilers/CSharp/Test/Emit2/Attributes/InternalsVisibleToAndStrongNameTests.cs:188]: Is there a corresponding test I should refer to where the signing temp path *is* available?

File: src/Compilers/Shared/BuildClient.cs (hunk 4 above)
Comment (jaredpar) [reply src/Compilers/Shared/BuildClient.cs:224]: yeah we should.

File: src/Compilers/Shared/BuildClient.cs (hunk 7 above)
Comment (jaredpar) [reply src/Compilers/Shared/BuildClient.cs:220]: I took @cston suggestion to add an error here. 

This is the kind of situation where it should never happen but technically it can if you really screw with environment variables so you have to handle that outcome.

File: src/Compilers/Shared/BuildServerConnection.cs (hunk 5 above)
Comment (jaredpar) [reply src/Compilers/Shared/BuildServerConnection.cs:16]: This feedback about unused usings was used.

Comment (jaredpar) [reply src/Compilers/Shared/BuildServerConnection.cs:16]: This feedback about unused usings was used.

File: src/Compilers/Shared/BuildServerConnection.cs (hunk 8 above)
Comment (jaredpar) [reply src/Compilers/Shared/BuildServerConnection.cs:609]: This PR originally started as trying to fix the temp dir which would include this variable but at the moment it does not. Let me fix this.

File: src/Compilers/VisualBasic/Test/Emit/Attributes/EmitTestStrongNameProvider.vb (hunk 9 above)
Comment (jaredpar) [reply src/Compilers/VisualBasic/Test/Emit/Attributes/EmitTestStrongNameProvider.vb:23]: Correct. `Temp` handles all the directories / paths created under it.

File: src/Compilers/Core/Portable/CommandLine/CommonCompiler.cs [hunk 12]
```diff
@@ -43,7 +43,7 @@ namespace Microsoft.CodeAnalysis
         internal string? SdkDirectory { get; }
 
         /// <summary>
-        /// The temporary directory a compilation should use instead of <see cref="Path.GetTempPath"/>.  The latter
+        /// The temporary directory a compilation should use instead of Path.GetTempPath.  The latter
```
[the remark this answers is not in the document]
Comment (jaredpar) [reply src/Compilers/Core/Portable/CommandLine/CommonCompiler.cs:46]: Correct. That is why I had to remove it.

File: src/Compilers/CSharp/Test/Emit/Emit/DesktopStrongNameProviderTests.cs (hunk 3 above)
Comment (jaredpar) [reply src/Compilers/CSharp/Test/Emit/Emit/DesktopStrongNameProviderTests.cs:28]: Changed to `RespectNullTempPath`

File: src/Compilers/CSharp/Test/Emit2/Attributes/InternalsVisibleToAndStrongNameTests.cs (hunk 10 above)
Comment (jaredpar) [reply src/Compilers/CSharp/Test/Emit2/Attributes/InternalsVisibleToAndStrongNameTests.cs:190]: removed the variable as I didn't see aneed for it anymore.

File: src/Compilers/CSharp/Test/Emit2/Attributes/InternalsVisibleToAndStrongNameTests.cs (hunk 11 above)
Comment (jaredpar) [reply src/Compilers/CSharp/Test/Emit2/Attributes/InternalsVisibleToAndStrongNameTests.cs:188]: There are but it's not as clear. Added one here that is much clearer opposite of this one.

File: src/Compilers/Core/Portable/CommandLine/CommonCompiler.CompilerEmitStreamProvider.cs [hunk 13]
```diff
@@ -45,8 +45,6 @@ public void Close(DiagnosticBag diagnostics)
                 }
             }
 
-            public override Stream? Stream => null;
```
[the remark this answers is not in the document]
Comment (jaredpar) [reply src/Compilers/Core/Portable/CommandLine/CommonCompiler.CompilerEmitStreamProvider.cs:48]: Yes. The virtual member was removed in the base type. 

The reason for this part of the change is I had a lot of time understanding the relationship between the `EmitStreamProvider` and underlying `Stream`. Particularly when do they get created and what is the ownership model. 
A significant part of this change was figuring out exactly where the files were created and destroyed hence as part of the fix I tried to make this much clearer. 

Happy for feedback on the new approach.

File: src/Compilers/Core/Portable/Compilation.EmitStream.cs (hunk 6 above)
Comment (jaredpar) [reply src/Compilers/Core/Portable/Compilation.EmitStream.cs:46]: Moved the comment to the more appropriate location.

File: .vscode/settings.json (hunk 2 above)
Comment (RikkiGibson) [inline .vscode/settings.json:1]: Is this deletion related to the goal of the PR?

**Review by RikkiGibson — APPROVED** [review]

**Comment by jaredpar** [comment]
@RikkiGibson FYI i pushed on more commit that had functional impact. In reviewing my change this morning and thinking about the followup change I realized I'd error'd in one place. Specifically not sending compilation requests to the server when the request cannot calculate a temporary directory. Originally I did that under the logic of the server can't start if there is a missing temp directory so what's the point. I now realize that was wrong for two reasons:

1. The temp directory calculated by the client and server can, and sometimes will be very different. Love or hate it builds change environment variables and that can change what value we get for a temp. Just because the client can't calculate a temp doesn't mean the server can't either. It could have been running from a previous request where environment variables were differente. 
2. A build request does not actually need a temporary directory in the 99.9% case. So treating them as invalid is wrong.
3. The responsibility was just in the wrong place. The server is the best to know if it can / can't startup. There is already loads of logic to deal with failed start up. The client should just send the request and let the server deal with startup conditions. 

Once I start on the next change I'm goingto add the fast fail code for the missing temp dir case. It's an extreme edge anyways.

**Comment by RikkiGibson** [comment]
Got it, so you just reverted the check to ensure client can get a temp directory before sending the build request? And in a subsequent PR you'll add some new check to ensure the server fails in a clear/actionable way when it can't create a temp directory?

File: src/Compilers/Core/Portable/Compilation.EmitStream.cs [hunk 14]
```diff
@@ -130,14 +131,30 @@ internal void Close()
...
+                    {
+                        diagnostics.Add(StrongNameKeys.GetError(
+                            _strongNameKeys.KeyFilePath,
+                            _strongNameKeys.KeyContainer,
+                            new CodeAnalysisResourcesLocalizableErrorArgument(nameof(CodeAnalysisResources.SigningTempPathUnavailable)),
+                            messageProvider));
+                        return null;
+                    }
+
+                    _stream = _emitStreamProvider.GetOrCreateStream(diagnostics);
```
Comment (cston) [inline src/Compilers/Core/Portable/Compilation.EmitStream.cs:146]: >_stream = _emitStreamProvider.GetOrCreateStream(diagnostics); [](http://example.com/codeflow?start=20&length=61)

It's not clear what the relationship is between this `_stream` and the `tempStream` we're returning below. We don't appear to be using `_stream` after creating it here.

File: src/Compilers/Shared/BuildServerConnection.cs [hunk 15]
```diff
@@ -31,12 +32,25 @@ internal sealed class BuildServerConnection
         /// </summary>
         internal static bool IsCompilerServerSupported => GetPipeName("") is object;
 
+        /// <summary>
+        /// Create a build request for processing on the server. 
+        /// </summary>
+        /// <remarks>
+        /// Even though compilation itself does not specifically require a temporary directory to be successful
+        /// this API deliberately requires <paramref name="tempDirectory"/> to have a value. The reason for this is
```
Comment (cston) [inline src/Compilers/Shared/BuildServerConnection.cs:40]: Is this still true after the most recent commit?

File: src/Compilers/Shared/BuildServerConnection.cs [hunk 16]
```diff
@@ -31,12 +31,25 @@ internal sealed class BuildServerConnection
...
+        /// hence won't be able to start. So creating <see cref="BuildRequest"/> instances isn't useful cause it's 
+        /// wasting time trying to start a server that won't be able to start.
+        /// </remarks>
         internal static BuildRequest CreateBuildRequest(
             Guid requestId,
             RequestLanguage language,
             List<string> arguments,
             string workingDirectory,
-            string tempDirectory,
+            string? tempDirectory,
```
Comment (cston) [inline src/Compilers/Shared/BuildServerConnection.cs:52]: >? [](http://example.com/codeflow?start=18&length=1)

Is the nullable annotation necessary given that the  method requires a value?

File: src/Compilers/Shared/BuildServerConnection.cs [hunk 17]
```diff
@@ -588,51 +601,63 @@ internal static string GetClientMutexName(string pipeName)
...
+                : getTempPathWindows(workingDir);
+
+            static string? getTempPathLinux()
             {
                 // Unix temp path is fine: it does not use the working directory
                 // (it uses ${TMPDIR} if set, otherwise, it returns /tmp)
+                //
+                // https://github.com/dotnet/roslyn/issues/65415 tracks moving to a directory under
+                // $XDG_RUNTIME_DIR if it is set using appropriate fall backs as necessary.
+                // https://specifications.freedesktop.org/basedir-spec/basedir-spec-latest.html
```
Comment (cston) [inline src/Compilers/Shared/BuildServerConnection.cs:622]: I don't understand this comment and given the issue referenced here, could we just say "to a per-user directory"?

File: src/Compilers/Shared/BuildServerConnection.cs (hunk 16 above)
Comment (jaredpar) [reply src/Compilers/Shared/BuildServerConnection.cs:52]: Good catch. The annotation is correct here, the `Debug.Assert` is wrong.

**Review by cston — APPROVED** [review]

File: src/Compilers/Shared/BuildServerConnection.cs (hunk 16 above)
Comment (Neme12) [reply src/Compilers/Shared/BuildServerConnection.cs:52]: The doc comment says that this API requires `tempDirectory` to have a value even though it's nullable. Is the comment wrong?

File: src/Compilers/Shared/BuildServerConnection.cs (hunk 15 above)
Comment (Neme12) [reply src/Compilers/Shared/BuildServerConnection.cs:39]: @jaredpar It looks like you didn't address this? This comment is still there.

Comment (jaredpar) [reply src/Compilers/Shared/BuildServerConnection.cs:39]: Had some issues where I didn't save before push and missed a few items. Cleaned most of them up. Looks like I missed this one. Will fix this in #69406.
